### PR TITLE
Reorganize code to allow conditional enablement of runtimes

### PR DESCRIPTION
**Pull request by dims**
Change-Id: I76583736d7ad39190a1a2bca820d4e957caadc84

### CODE REVIEW

**Comment by dashpole** [comment]
Let me know when this is ready for review

**Comment by dims** [comment]
@dashpole this is ready now (just rebased)

**Comment by dims** [comment]
cc @tallclair

**Comment by tallclair** [comment]
This is awesome (I haven't thoroughly reviewed it yet though)

What is the plan for integrating this into Kubernetes though?

BTW, there are still a bunch of docker dependencies, but those probably deserve to be addressed independently.
image

File: container/mesos/init.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/watcher"
+	"k8s.io/klog"
+)
+
+func init() {
+	err := container.RegisterPlugin("mesos", func(factory info.MachineInfoFactory, fsInfo fs.FsInfo, includedMetrics container.MetricSet) (watcher.ContainerWatcher, error) {
+		err := Register(factory, fsInfo, includedMetrics)
+		if err != nil {
+			klog.V(5).Infof("Registration of the mesos container factory failed: %v.", err)
```
Comment (dashpole) [inline container/mesos/init.go]: Aren't all of these logs a duplicate of the one in InitializePlugins?  I think we could just `return nil, err` here, and in other factories.

Comment (dims) [reply container/mesos/init.go]: Fixed. thanks!

**Comment by dims** [comment]
@tallclair yep! wanted to peel off some of the leaves first (like mesos)

In kubernetes, we will import only the items corresponding to the CRI(s) that we support.

File: container/containerd/init.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/google/cadvisor/fs"
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/watcher"
+	"k8s.io/klog"
+)
+
+func init() {
+	err := container.RegisterPlugin("containerd", func(factory info.MachineInfoFactory, fsInfo fs.FsInfo, includedMetrics container.MetricSet) (watcher.ContainerWatcher, error) {
+		err := Register(factory, fsInfo, includedMetrics)
+		if err != nil {
```
Comment (dashpole) [inline container/containerd/init.go]: We don't need to check err here or in other plugins that don't register any watchers.  Just `return nil, err`

Comment (dims) [reply container/containerd/init.go]: Done!

**Pull request merged**